### PR TITLE
[CDAP-3516] Added a MetadataService that runs in the dataset executor…

### PR DESCRIPTION
**Pull request by bdmogal**
… container. Moved metadata HTTP APIs to this service from Dataset Service.

Also moved some Metadata classes to cdap-app-fabric.

### CODE REVIEW

**Comment by bdmogal** [comment]
Build: http://builds.cask.co/browse/CDAP-DUT2786
Jira: [CDAP-3516](https://issues.cask.co/browse/CDAP-3516)

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataServiceManager.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+/**
+ * Service manager for {@link MetadataService} in distributed mode.
+ */
+public class MetadataServiceManager extends AbstractDistributedMasterServiceManager {
+
+  @Inject
+  MetadataServiceManager(CConfiguration cConf, TwillRunnerService twillRunnerService,
+                                DiscoveryServiceClient discoveryServiceClient) {
```
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataServiceManager.java]: small nit mis-align parameters

**Comment by henry-cask** [comment]
LGTM pending test on cluster

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+
+  @BeforeClass
+  public static void setup() throws MalformedURLException {
+    DiscoveryServiceClient discoveryClient = getInjector().getInstance(DiscoveryServiceClient.class);
+    ServiceDiscovered metadataHttpDiscovered = discoveryClient.discover(Constants.Service.METADATA_SERVICE);
+    EndpointStrategy endpointStrategy = new RandomEndpointStrategy(metadataHttpDiscovered);
+    Discoverable discoverable = endpointStrategy.pick(1, TimeUnit.SECONDS);
+    Assert.assertNotNull(discoverable);
+    int port = discoverable.getSocketAddress().getPort();
+    metadataServiceUrl = String.format("http://127.0.0.1:%d/", port);
```
Comment (anwar6953) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: extra slash at the end of metadataServiceUrl

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+    HttpResponse response = makeGetRequest(path);
+    Assert.assertEquals(200, response.getResponseCode());
+    return GSON.fromJson(response.getResponseBodyAsString(), LIST_STRING_TYPE);
+  }
+
+  private HttpResponse makePostRequest(String resource) throws IOException {
+    return makePostRequest(resource, null);
+  }
+
+  private HttpResponse makePostRequest(String resource, @Nullable String body) throws IOException {
```
Comment (anwar6953) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: there's already a doPost method available here, right?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+  private List<String> getTags(Id.DatasetInstance dataset) throws IOException {
+    String path = getVersionedAPIPath(String.format("datasets/%s/tags", dataset.getId()), dataset.getNamespaceId());
+    HttpResponse response = makeGetRequest(path);
+    Assert.assertEquals(200, response.getResponseCode());
+    return GSON.fromJson(response.getResponseBodyAsString(), LIST_STRING_TYPE);
+  }
+
+  private List<String> getTags(Id.Stream stream) throws IOException {
+    String path = getVersionedAPIPath(String.format("streams/%s/tags", stream.getId()), stream.getNamespaceId());
+    HttpResponse response = makeGetRequest(path);
```
Comment (anwar6953) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: there's already a doGet method available here, right?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java (hunk 3 above)
Comment (anwar6953) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: At least comment why we need another makePostRequest method.

**Comment by poornachandra** [comment]
Since we don't expose the service in system services list, I think we should route ping to metadata service. So that we can figure out if the service is running or not in production.

**Comment by anwar6953** [comment]
Why not expose it in system services list?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataService.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  private final MetricsCollectionService metricsCollectionService;
+  private final DiscoveryService discoveryService;
+  private final Set<HttpHandler> handlers;
+
+  private NettyHttpService httpService;
+
+  @Inject
+  MetadataService(CConfiguration cConf, MetricsCollectionService metricsCollectionService,
+                  DiscoveryService discoveryService,
+                  @Named("metadata.handlers") Set<HttpHandler> handlers) {
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataService.java]: It is good to make `metadata.handlers` a constant.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java [hunk 6]
```diff
@@ -63,7 +63,7 @@ public void testSystemServices() throws Exception {
     List<SystemServiceMeta> actual = GSON.fromJson(new String(ByteStreams.toByteArray(urlConn.getInputStream()),
                                                        Charsets.UTF_8), token);
 
-    Assert.assertEquals(7, actual.size());
+    Assert.assertEquals(8, actual.size());
```
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: This should go back to 7, I think.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java [hunk 7]
```diff
@@ -78,7 +78,7 @@ public void testSystemServicesStatus() throws Exception {
 
     Map<String, String> result = GSON.fromJson(new String(ByteStreams.toByteArray(urlConn.getInputStream()),
                                                Charsets.UTF_8), token);
-    Assert.assertEquals(7, result.size());
+    Assert.assertEquals(8, result.size());
```
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: Same here.

**Comment by bdmogal** [comment]
@anwar6953 I have added https://issues.cask.co/browse/CDAP-3635 to handle management/monitoring. It requires some more thinking. Will do it later.

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java (hunk 4 above)
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: fixed

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java (hunk 3 above)
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: added comment

File: cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java [hunk 8]
```diff
@@ -231,6 +235,7 @@ public void shutDown() {
         externalAuthenticationServer.stopAndWait();
       }
       logAppenderInitializer.close();
+      metadataService.startAndWait();
```
Comment (poornachandra) [inline cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java]: This should be `stopAndWait`

Comment (bdmogal) [inline cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java]: oh gosh! nice catch. fixed.

**Comment by bdmogal** [comment]
Build is green. @poornachandra, feel free to merge this after taking a final look. I will take care of https://github.com/caskdata/cdap/pull/3868#issuecomment-139427272 in a separate PR, since I want to get this in quicker.

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+    EndpointStrategy endpointStrategy = new RandomEndpointStrategy(metadataHttpDiscovered);
+    Discoverable discoverable = endpointStrategy.pick(1, TimeUnit.SECONDS);
+    Assert.assertNotNull(discoverable);
+    int port = discoverable.getSocketAddress().getPort();
+    metadataServiceUrl = String.format("http://127.0.0.1:%d", port);
+  }
+
+  @Test
+  public void testMetadata() throws Exception {
+    //non-existing namespace
```
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: What does this mean?

**Comment by poornachandra** [comment]
@bdmogal One trivial comment, please squash and merge.

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java (hunk 9 above)
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: misplaced comment. was a placeholder, but verifications for non-existing namespace are at the end of this test. will remove.

**Comment by bdmogal** [comment]
Squashed commits. merging.

**Pull request merged**